doc/developer-notes.md

```diff
@@ -839,6 +840,36 @@ namespace {
 #endif // BITCOIN_FOO_BAR_H
 ```
 
+Construct On First Use
+------------------------
+
+It is not possible to control the order in which non-local static objects in *different* translation units are initialized.
+This is the so-called "static initialization order problem". It could cause the use of an object:
+- before it is constructed (on program startup)
+- after it is destructed (at program exit)
+
+Both cases lead to undefined behavior.
+
+To prevent the static initialization order problem, be sure to Construct On First Use:
+
+```C++
+BCLog::Logger& LogInstance()
+{
+    // See doc/developer-notes.md#construct-on-first-use
+    static BCLog::Logger* g_logger{new BCLog::Logger()};
```

Comment:
👀 
```cpp
static BCLog::Logger logger;
return logger;
```

Reply:
> ```c++
> static BCLog::Logger logger;
> return logger;
> ```

The `logger` object could (potentially) be destroyed before its last use at program exit sequence that leads to undefined behavior.

Reply:
Actually, the d954f7c8c017c7e9bf60f48a8c41da8c498239f7 commit in this PR fixes such issue for the `lockdata` object.

Reply:
Changes to the `developer-notes.md` [dropped](https://github.com/bitcoin/bitcoin/pull/18881#issuecomment-625296903).